File: src/main/java/com/github/frcsty/frozenactions/time/TimeScanner.java
package com.github.frcsty.frozenactions.time;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.function.Predicate;

@SuppressWarnings("StatementWithEmptyBody")
final class TimeScanner {

    private final char[] time;
    private int index = 0;

    TimeScanner(final @NotNull String time) {
        this.time = time.toCharArray();
    }

    boolean hasNext() {
        return index < time.length - 1;
    }

    long nextLong() {
        return Long.parseLong(String.valueOf(next(Character::isDigit)));
    }

    @NotNull String nextString() {
        return String.copyValueOf(next(codePoint ->
                Character.isAlphabetic(Integer.valueOf(codePoint))
        ));
    }

    private char[] next(final @NotNull Predicate<Character> condition) {
        final int startIndex = index;
        while (++index < time.length && condition.test(time[index])) ;
        return Arrays.copyOfRange(time, startIndex, index);
    }

}

File: src/main/kotlin/com/github/frcsty/frozenactions/time/TimeScanner.kt
package com.github.frcsty.frozenactions.time

internal class TimeScanner(time: String) {
    private val time = time.toCharArray()
    private var index = 0

    operator fun hasNext(): Boolean {
        return index < time.size - 1
    }

    fun nextLong(): Long {
        return String(next(Char::isDigit)).toLong()
    }

    fun nextString(): String {
        return String(next { codePoint ->
            Character.isAlphabetic(codePoint.toInt())
        })
    }

    private fun next(condition: (Char) -> Boolean): CharArray {
        val startIndex = index
        while (++index < time.size && condition(time[index]));
        return time.copyOfRange(startIndex, index)
    }

}